### PR TITLE
acquireLogin/releaseLogin apis for data controller to ensure only one dc is logged in at a given time

### LINKED ISSUE #13875: ControllerModel needs to provide synchronization apis to ensure azdata login does not change abruptly.
Currently, execution of various SQL mi and postgres operations like edit and delete issue azdataLogin before doing the main operation to ensure that operation is done against the correct data controller. However, two different executions of these commands could trample on each other.

There needs to be a synchronization mechanism provided around AzdataLogin apis so that a new operation about to start can be chained to start when the previous one finishes..

_Originally posted by @ranasaria in https://github.com/microsoft/azuredatastudio/pull/13569#r548423654_

### PR DESCRIPTION
**Pull request by ranasaria**
This change adds new api on arc ControllerModel object called acuireLogin and releaseLogin which ensure that a new login is not performed until previous logged in session has been released.

This PR fixes #13875

@charles-gagnon is taking over completion of this PR as  needed.

### CODE REVIEW

File: extensions/arc/src/models/controllerModel.ts [hunk 1]
```diff
@@ -60,11 +60,36 @@ export class ControllerModel {
...
+	 * then this call waits for the previous login session to finish before acquiring the login session.
+	 * @param promptReconnect - if true then the user is prompted for connection information.
+	 */
+	public async acquireLogin(promptReconnect: boolean = false): Promise<void> {
+		//wait for any previous login sessions that might be in progress to finish
+		if (ControllerModel._loginSessionInProgress) {
+			await ControllerModel._loginSessionInProgress;
+		}
+		// create a new in progress promise object
+		ControllerModel._loginSessionInProgress = new Deferred<void>();
```
Comment (Charles-Gagnon) [inline extensions/arc/src/models/controllerModel.ts:74]: If multiple callers are waiting on this then they'll all get notified and continue execution at once which is the exact problem you're trying to fix in the first place.

File: extensions/arc/src/models/controllerModel.ts [hunk 2]
```diff
@@ -60,11 +60,36 @@ export class ControllerModel {
 		};
 	}
 
+	/**
+	 * acquires a login session to this controller. If a login session is in progress against any controller
+	 * then this call waits for the previous login session to finish before acquiring the login session.
+	 * @param promptReconnect - if true then the user is prompted for connection information.
+	 */
+	public async acquireLogin(promptReconnect: boolean = false): Promise<void> {
```
Comment (Charles-Gagnon) [inline extensions/arc/src/models/controllerModel.ts:68]: What about other callers to azdata calling login? Both from other extensions in ADS and from outside?

We need to decide how much effort we put into preventing this issue - the reality though currently is that we'll never be able to be 100% sure that we aren't changing the login context until we get the context parameter support we've previously discussed. 

What immediate concerns do you have with our current approach such that these changes are necessary now? Currently the code does a login and then immediately runs the command which should avoid most issues unless people are doing multiple actions simultaneously.

File: extensions/arc/src/ui/dashboards/postgres/postgresOverviewPage.ts [hunk 3]
```diff
@@ -188,7 +192,7 @@ export class PostgresOverviewPage extends DashboardPage {
 								cancellable: false
 							},
 							async (_progress, _token) => {
-								await this._postgresModel.controllerModel.azdataLogin();
+								await this._postgresModel.controllerModel.acquireLogin();
```
Comment (Charles-Gagnon) [inline extensions/arc/src/ui/dashboards/postgres/postgresOverviewPage.ts:195]: You never release the login

**Review by Charles-Gagnon — CHANGES_REQUESTED** [review]

File: extensions/arc/src/models/controllerModel.ts (hunk 2 above)
Comment (ranasaria) [reply extensions/arc/src/models/controllerModel.ts:68]: While the arc controllers' tree is refreshing if you open management the management page of miaa/pg for a controller that has refreshed then results are unpredictable for the arc controller that is yet to refresh.

For other callers to azdata login from other extensions, they would need to take care of this synchronization themselves. In our current codebase, what other similar call are you worried about?

File: extensions/arc/src/models/controllerModel.ts (hunk 1 above)
Comment (ranasaria) [reply extensions/arc/src/models/controllerModel.ts:74]: hmmm - let me think more on that. Do you have a suggestion?

Comment (ranasaria) [reply extensions/arc/src/models/controllerModel.ts:74]: I am looking into chaining all the pending callers so they go in a sequence

File: extensions/arc/src/models/controllerModel.ts (hunk 2 above)
Comment (Charles-Gagnon) [reply extensions/arc/src/models/controllerModel.ts:68]: Is this something that you've seen while testing? I'm surprised that it'd be that big of issue given the nature of the call sequencing as it is. 

Let's talk this through more before you spend more time on this - it doesn't need to be fixed immediately and I want to make sure we're approaching this correctly.

Comment (ranasaria) [reply extensions/arc/src/models/controllerModel.ts:68]: Yes i have observed this in testing and is an easy repro.

**Pull request closed without merging**